File: baitaptuan3/Julia/bai1.jl
# Hàm lấy các vị trí lân cận (trên, dưới, trái, phải)
function get_neighbors(grid, position)
    x, y = position
    neighbors = []
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Trên, Dưới, Trái, Phải

    for (dx, dy) in directions
        nx, ny = x + dx, y + dy
        if nx >= 1 && nx <= size(grid, 1) && ny >= 1 && ny <= size(grid, 2)
            if grid[nx, ny] != 1  # 1 là chướng ngại vật
                push!(neighbors, (nx, ny))
            end
        end
    end
    return neighbors
end

# Hàm DFS tìm đường đi từ start đến goal
function dfs(grid, start, goal)
    stack = [start]
    visited = Set()
    parent = Dict{Tuple{Int, Int}, Tuple{Int, Int}}()

    while !isempty(stack)
        current = pop!(stack)

        if current == goal
            break
        end

        if current in visited
            continue
        end

        push!(visited, current)

        # Lấy các vị trí lân cận
        neighbors = get_neighbors(grid, current)
        for neighbor in neighbors
            if !(neighbor in visited)
                push!(stack, neighbor)
                parent[neighbor] = current
            end
        end
    end

    # Kiểm tra đường đi
    path = []
    if !(haskey(parent, goal)) && start != goal
        println("Không tìm thấy đường đi.")
        return path
    end

    # Xây dựng đường đi từ goal về start
    current = goal
    while current != start
        push!(path, current)
        current = get(parent, current, start)
    end
    push!(path, start)

    return reverse(path)
end

# Minh họa
grid = [
    [0 0 0 0 0],
    [1 1 0 1 0],
    [0 0 0 1 0],
    [0 1 1 1 0],
    [0 0 0 0 0]
]

start = (1, 1)
goal = (5, 5)
path = dfs(grid, start, goal)

if !isempty(path)
    println("Đường đi tìm được:")
    for p in path
        println(p)
    end
end

File: Buoi 3/C++/bai1.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_set>
#include <map>
#include <utility>
#include <algorithm>

using namespace std;

// Hàm lấy các vị trí lân cận (trên, dưới, trái, phải)
vector<pair<int, int>> get_neighbors(const vector<vector<int>>& grid, pair<int, int> position) {
    int x = position.first, y = position.second;
    vector<pair<int, int>> neighbors;
    vector<pair<int, int>> directions = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };  // Trên, Dưới, Trái, Phải

    for (auto& dir : directions) {
        int nx = x + dir.first, ny = y + dir.second;
        if (nx >= 0 && nx < grid.size() && ny >= 0 && ny < grid[0].size()) {
            if (grid[nx][ny] != 1) {  // 1 là chướng ngại vật
                neighbors.push_back({nx, ny});
            }
        }
    }
    return neighbors;
}

// Hàm DFS tìm đường đi từ start đến goal
vector<pair<int, int>> dfs(const vector<vector<int>>& grid, pair<int, int> start, pair<int, int> goal) {
    stack<pair<int, int>> s;
    s.push(start);
    unordered_set<string> visited;
    map<pair<int, int>, pair<int, int>> parent;

    while (!s.empty()) {
        pair<int, int> current = s.top();
        s.pop();

        if (current == goal) {
            break;
        }

        if (visited.find(to_string(current.first) + "," + to_string(current.second)) != visited.end()) {
            continue;
        }

        visited.insert(to_string(current.first) + "," + to_string(current.second));

        // Lấy các vị trí lân cận
        vector<pair<int, int>> neighbors = get_neighbors(grid, current);
        for (auto& neighbor : neighbors) {
            if (visited.find(to_string(neighbor.first) + "," + to_string(neighbor.second)) == visited.end()) {
                s.push(neighbor);
                parent[neighbor] = current;
            }
        }
    }

    // Kiểm tra đường đi
    vector<pair<int, int>> path;
    if (parent.find(goal) == parent.end() && start != goal) {
        cout << "Khong tim thay duong didi." << endl;
        return path;
    }

    // Xây dựng đường đi từ goal về start
    pair<int, int> current = goal;
    while (current != start) {
        path.push_back(current);
        current = parent[current];
    }
    path.push_back(start);
    reverse(path.begin(), path.end());

    return path;
}

int main() {
    vector<vector<int>> grid = {
        {0, 0, 0, 0, 0},
        {1, 1, 0, 1, 0},
        {0, 0, 0, 1, 0},
        {0, 1, 1, 1, 0},
        {0, 0, 0, 0, 0}
    };

    pair<int, int> start = {0, 0};
    pair<int, int> goal = {4, 4};
    vector<pair<int, int>> path = dfs(grid, start, goal);

    if (!path.empty()) {
        cout << "Duong di tim duocduoc:" << endl;
        for (auto& p : path) {
            // Sửa lỗi tại đây để in dấu phẩy và ngoặc đơn đúng cách
            cout << "(" << p.first << ", " << p.second << ") ";
        }
        cout << endl;
    } else {
        cout << "Khong tim thay duong didi." << endl;
    }

    return 0;
}
